types/rdf-js/index.d.ts

```diff
@@ -476,3 +476,181 @@ export interface DatasetCoreFactory<Q extends BaseQuad = Quad> {
      */
     dataset(quads?: Q[]): DatasetCore<Q>;
 }
+
+export interface Dataset<Q extends BaseQuad = Quad> extends DatasetCore<Q> {
+    match(subject?: Term, predicate?: Term, object?: Term, graph?: Term): Dataset<Q>;
```

Comment:
Not aware of a way to have the return type on `DatasetCore.match` be the same as the current type (`this` means the same instance?).

Reply:
`Dataset<Q>` should be fine.

Reply:
Reading https://www.typescriptlang.org/docs/handbook/advanced-types.html#polymorphic-this-types, `this` doesn't actually seem to mean the same instance. So can remove this and update `DatasetCore.match` to return `this`.

Reply:
Though the return types on https://rdf.js.org/dataset-spec/#datasetcore-interface is confusing. Took `add`/`delete` to mean it returns the same instance (so `DatasetCore`, `Dataset`, or some other subtype), and `match` to return a new instance of the same type (even though the IDL says `Dataset` and the description 'This method always returns a new `DatasetCore`').

Reply:
Done in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40794/commits/5b643c79c97291f68f4048091be841e438058897 (could still use confirming that there's no way to separate the two cases).